flow/enhancers/socket_info.go

```diff
@@ -0,0 +1,395 @@
+	s.resetScanProcNet()
+
+	pids := getPIDs("/proc")
+	for _, p := range pids {
+		procTask := "/proc/" + p + "/task"
+		tasks := getPIDs(procTask)
+		for _, t := range tasks {
+			process := procTask + "/" + t
+			s.updateProcNet(process)
+
+			fds := getFDs(process)
+			for _, f := range fds {
+				inode := rFdSocket.FindStringSubmatch(f)
+				if len(inode) == 0 {
+					continue
+				}
+
+				var isUDP = true
+				sinfo, ok := s.cacheUDP.inode.Get(inode[1])
+				if !ok {
+					isUDP = false
+					sinfo, ok = s.cacheTCP.inode.Get(inode[1])
+					if !ok {
+						continue
+					}
+				}
+				si := sinfo.(SocketInfo)
+				parseProcStat(process, &si)
+				if isUDP == true {
+					s.cacheUDP.inode.Set(inode[1], si, cache.DefaultExpiration)
+				} else {
+					s.cacheTCP.inode.Set(inode[1], si, cache.DefaultExpiration)
+				}
+			}
+		}
+	}
+}
+
+func ipToHex(ip46 net.IP) string {
+	s := ""
+	ip := ip46.To4()
+	if ip == nil {
+		ip = ip46
+	}
+	for i := range ip {
+		s += fmt.Sprintf("%02X", ip[(len(ip)-1)-i])
+	}
+	return s
+}
+
+func (s *SocketInfoEnhancer) getSocketInfoLocal(f *flow.Flow) bool {
+	ip := net.ParseIP(f.Network.A)
+	if ip == nil {
+		return false
+	}
+
+	port, _ := strconv.ParseUint(f.Transport.A, 10, 0)
+	ipPortA := fmt.Sprintf("%s:%04X", ipToHex(ip), port)
+	ipPortB := ""
+
+	ip = net.ParseIP(f.Network.B)
+	if ip != nil {
+		port, _ = strconv.ParseUint(f.Transport.B, 10, 0)
+		ipPortB = fmt.Sprintf("%s:%04X", ipToHex(ip), port)
+	}
+
+	var cache *socketCache
+	if f.Transport.Protocol == flow.FlowProtocol_UDPPORT {
+		cache = s.cacheUDP
+	} else if f.Transport.Protocol == flow.FlowProtocol_TCPPORT {
+		cache = s.cacheTCP
+	}
+
+	retried := false
+retry:
+	var siA, siB bool
+	// si.LocalIPPort == ipPortA
+	if local, ok := cache.localAddr.Get(ipPortA); ok {
+		if ca, ok := local.(cacheAddr); ok {
+			if sinfo, ok := cache.inode.Get(ca.inode); ok {
+				if si, ok := sinfo.(SocketInfo); ok {
+					f.SocketA = SocketInfoToFlowSocketInfo(&si)
+					siA = true
+				}
+			}
+		}
+	}
+	// peer (si.LocalIPPort == ipPortB) && (si.RemoteIPPort == ipPortA)
+	if local, ok := cache.localRemoteAddr.Get(ipPortB + ipPortA); ok {
+		if ca, ok := local.(cacheAddr); ok {
+			if sinfo, ok := cache.inode.Get(ca.inode); ok {
+				if si, ok := sinfo.(SocketInfo); ok {
+					f.SocketB = SocketInfoToFlowSocketInfo(&si)
+					siB = true
+				}
+			}
+		}
+	}
+	if siA || siB {
+		return true
+	}
+
+	if retried == false {
+		s.scanProc()
+		retried = true
+		goto retry
+	}
+	return false
+}
+
+// SocketInfoToFlowSocketInfo create a new flow.SocketInfo based on process SocketInfo
+func SocketInfoToFlowSocketInfo(si *SocketInfo) *flow.SocketInfo {
+	if si == nil {
+		return nil
+	}
+	return &flow.SocketInfo{
+		Process: si.Exe,
+		Pid:     int64(si.Pid),
+		Ppid:    int64(si.PPid),
+		Name:    si.Name,
+		Tid:     int64(si.Tid),
+	}
+}
+
+func (s *SocketInfoEnhancer) getFlowSocketInfoLocal(f *flow.Flow) {
+	if foundSI := s.getSocketInfoLocal(f); foundSI == false {
```

Comment:
my concern here is if we start a capture on a node that doesn't know any of A or B we will try for every packet. What about trying then skipping whatever the result ?